firmware/ventilator-controller-stm32/Core/Test/Pufferfish/Protocols/CRCElements.cpp

```diff
@@ -0,0 +1,151 @@
+      TestCRCElement crc_element{payload};
+
+      PF::Util::ByteVector<buffer_size> input_buffer;
+      PF::Util::convertStringToByteVector(body, input_buffer);
+
+      auto parse_status = crc_element.parse(input_buffer);
+      auto crc_check = crc_element.crc();
+      auto payload_check = crc_element.payload();
+      THEN("the final status should be ok") {
+        REQUIRE(parse_status == PF::IndexStatus::ok);
+        REQUIRE(crc_check != expected_crc);
+        REQUIRE(payload_check == expected_payload);
+      }
+    }
+
+    WHEN("data is written to it and parsed back") {
+      auto crc = PF::Util::make_array<uint8_t>(0x98, 0xdb, 0xe3, 0x55);
+      auto body = std::string("\x98\xdb\xe3\x55\x01\x05\x01\x02\x03\x04\x05", 11);
+
+      TestCRCElementProps::PayloadBuffer payload;
+      TestCRCElement crc_element{payload};
+
+      PF::Util::ByteVector<buffer_size> input_buffer;
+      PF::Util::convertStringToByteVector(body, input_buffer);
+
+      auto write_status = crc_element.write(input_buffer, crc32c);
+      auto parse_status = crc_element.parse(input_buffer);
+      THEN("the final statuses should be ok") {
+        REQUIRE(write_status == PF::IndexStatus::ok);
+        REQUIRE(parse_status == PF::IndexStatus::ok);
+      }
+    }
+  }
+
+  GIVEN("A CRC element receiver") {
+    constexpr size_t buffer_size = 254UL;
+    using TestCRCElementProps = PF::Protocols::CRCElementProps<buffer_size>;
+    using TestCRCElement = PF::Protocols::CRCElement<TestCRCElementProps::PayloadBuffer>;
+    PF::Util::ByteVector<buffer_size> input_buffer;
+
+    PF::HAL::SoftCRC32 crc32c{PF::HAL::crc32c_params};
+    PF::Protocols::CRCElementReceiver<buffer_size> crc_element_receiver{crc32c};
+
+    WHEN("good data is given as input to receiver") {
+      auto body = std::string("\x98\xdb\xe3\x55\x01\x05\x01\x02\x03\x04\x05", 11);
+
+      TestCRCElementProps::PayloadBuffer payload;
+      TestCRCElement crc_element{payload};
+
+      PF::Util::ByteVector<buffer_size> input_buffer;
+      PF::Util::convertStringToByteVector(body, input_buffer);
+
+      auto write_status = crc_element.write(input_buffer, crc32c);
+      auto status = crc_element_receiver.transform(input_buffer, crc_element);
+      THEN("the final status should be ok") {
+        REQUIRE(write_status == PF::IndexStatus::ok);
+        REQUIRE(status == PF::Protocols::CRCElementReceiver<buffer_size>::Status::ok);
+      }
+    }
+
+    WHEN("bad data is given as input to receiver") {
+      // this is not tested as body for bad data is good!
+    }
+  }
+
+  GIVEN("A CRC element sender") {
+    constexpr size_t buffer_size = 254UL;
+    using TestCRCElementProps = PF::Protocols::CRCElementProps<buffer_size>;
+    using TestCRCElement = PF::Protocols::CRCElement<TestCRCElementProps::PayloadBuffer>;
+
+    PF::HAL::SoftCRC32 crc32c{PF::HAL::crc32c_params};
+    PF::Protocols::CRCElementSender<buffer_size> crc_element_sender{crc32c};
+
+    WHEN("data is written to it") {
```

Comment:
This should be `when data of a valid length is given as input to the sender`. We also need tests for data just under, at, and just above the valid length limit - in some of those cases, a `Status::invalid_length` error code is supposed to be returned by the `transform` method.

Reply:
I had a doubt here, do we need to pass body ( crc + payload ) as the input_buffer or just the payload.
as the transform function calculated crc for the whole body and added it as prefix 
input_buffer = \x98\xdb\xe3\x55\x01\x05\x01\x02\x03\x04\x05
output_buffer = \x39\xEB\xD2\x48\x98\xDB\xE3\x55\x01\x05\x01\x02\x03\x04\x05

Reply:
Will add test cases for invalid_length.

Reply:
Yes, that sounds correct.